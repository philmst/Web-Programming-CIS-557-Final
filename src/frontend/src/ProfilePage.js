import './ProfilePage.css';
import React from 'react';

function ProfilePage() {
  return (
    <div className="App">
      <header className="App-header" />
      <body>
        <div id="cover_space">
          <img id="user_profile_pic" src="Stacy meditating.jpg" alt="" />
          <div id="main_title">Stacy Shapiro&apos;s FaceTok Page</div>
        </div>
        <div id="nav_button_container">
          <div className="nav_button">Friends</div>
          <div className="nav_button">Photos</div>
          <div className="nav_button">Groups</div>
          <div className="nav_button">Update bio</div>
          <div className="nav_button">Settings</div>
        </div>
        <div id="main_content_container">
          <div id="key_bio_info">
            <b> Stacy&apos;s bio </b>
            <p>
              I&apos;m just a wild eat pray love gal who also enjoys a
              bit of nuclear physics on the side!
            </p>
            <p> Penn: Class of 2019 </p>
            <p> NASA: 2020 - 2021 </p>
            <p> Chardy&apos;s with the gals: FOREVER!! </p>
          </div>
          <div id="post_container">
            <div className="wall_post">
              <div className="post_content_container">
                <b className="post_content_name"> Yide Zhao </b>
                <p className="post_content_message">OMG Stacy I can&apos;t believe you did that! You&apos;re so out there! #girlboss</p>
                <div className="reply_container">
                  <div className="post_reply_textbox"> Type your reply here...</div>
                  <div className="post_reply_button">Reply!</div>
                </div>
              </div>
            </div>
            <div className="wall_post">
              <div className="post_content_container">
                <b className="post_content_name"> Charlotte Dagabond </b>
                <p className="post_content_message">Stacy I hope you survived last night! </p>
                <div className="reply_container">
                  <div className="post_reply_textbox"> Type your reply here...</div>
                  <div className="post_reply_button">Reply!</div>
                </div>
              </div>
            </div>
            <div className="wall_post">
              <div className="post_content_container">
                <b className="post_content_name"> Bruce Yang </b>
                <p className="post_content_message">I&apos;m still cleaning up those vomit stains. I&apos;ll venmo you when I know the damage. </p>
                <div className="reply_container">
                  <div className="post_reply_textbox"> Type your reply here...</div>
                  <div className="post_reply_button">Reply!</div>
                </div>
              </div>
            </div>
            <div className="wall_post">
              <div className="post_content_container">
                <b className="post_content_name"> Professor Eric </b>
                <p className="post_content_message">Have you seen D&apos;Amelio&apos;s latest OOTD?</p>
                <div className="reply_container">
                  <div className="post_reply_textbox"> Type your reply here...</div>
                  <div className="post_reply_button">Reply!</div>
                </div>
              </div>
            </div>
            <div className="wall_post">
              <div className="post_content_container">
                <b className="post_content_name"> Turing Von Neumann </b>
                <p className="post_content_message">I&apos;m still shocked that you managed to prove Goldbach&apos;s conjecture on a napkin 8 margs in!</p>
                <div className="reply_container">
                  <div className="post_reply_textbox"> Type your reply here...</div>
                  <div className="post_reply_button">Reply!</div>
                </div>
              </div>
              <p />
            </div>
          </div>
        </div>
      </body>
    </div>
  );
}

export default ProfilePage;
