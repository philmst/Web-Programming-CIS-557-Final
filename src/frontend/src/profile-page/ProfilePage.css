#cover_space {
  min-height: 20vh;
  background-color: rgba(211, 211, 211, 0.2);
  padding-left: 3vh;
  padding-top: 3vh;
  padding-bottom: 3vh;
  margin-left: 20vw;
  margin-right: 20vw;
  text-align: center;
  font-weight: 600;
  font-size: 40px;
}

#main_title {
  line-height: 180px;
  height: 180px;
}

#user_profile_pic {
  height: 19vh;
  width: 19vh;
  background-color: lightblue;
  border-radius: 100%;
  justify-content: center;
  align-items: center;
  text-align: center;
  display: flex;
  font-weight: 500;
  float: left;
}

#nav_button_container {
  display: flex;
  justify-content: space-between;
  margin-left: 20vw;
  margin-right: 20vw;
  margin-top: 3vh;
  margin-bottom: 3vh;
  text-align: center;

}

.nav_button {
  background-color: lightblue;
  width: 5vw;
  color: black;
  padding-left: 3vw;
  padding-right: 3vw;
  padding-top: 3vh;
  padding-bottom: 3vh;
  font-weight: 500;
}

#main_content_container {
  
  margin-left: 20vw;
  margin-right: 20vw;
}

#key_bio_info {
  width: 15vw;
  float: left;
  
  padding: 3vh;
  height: 50%;
  background-color: rgba(211, 211, 211, 0.2);
}

#post_container {
  width: 50%;
  height: 50%;
  float: left;
  background-color: white;
}

.wall_post {
  margin-left: 3vh;
  margin-right: 3vh;
  margin-bottom: 3vh;
  background-color: rgba(211, 211, 211, 0.2);
  width: 37vw;
  padding-left: 3vh;
  padding-top: 3vh;
  padding-right: 3vh;
  padding-bottom: 2vh;

  display: block;
}

.main_post {
  font-size: 20px;
}

.post_caption {
  font-size: 20px;
}

.reply_content_name {
  font-size: 14px;
}

.reply_text {
  font-size: 14px;
}

/**.post_profile_pic_container {
  height: 50%;
  float: left;
  background-color: lightblue;
  border-radius: 100%;
  justify-content: center;
  align-items: center;
  text-align: center;
  display: block;
}

.post_profile_pic {
  height: 100%;
  width: 100%;
  float: center;
  background-color: lightblue;
  border-radius: 100%;
  justify-content: center;
  align-items: center;
  text-align: center;
  display: block;
}

.post_content_container {
  height: 50%;
  float: left;
  display: block;
} */

.reply_container {
  display: flex;
  justify-content: space-between;
}

.post_reply_textbox {
  border-width: 2px;
  border-color: grey;
  border-style: solid;
  width: 80%;
  padding: 3px;
  background-color: white;
}

.post_reply_button {
  border-width: 2px;
  border-color: rgb(157, 200, 214);
  border-style: solid;
  width: 15%;
  padding: 3px;
  background-color: lightblue;;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
