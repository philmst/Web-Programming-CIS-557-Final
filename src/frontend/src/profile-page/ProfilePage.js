import './ProfilePage.css';
import React, { useEffect, useState } from 'react';
// import { getProfile, getUserPosts } from './ProfileModule';
// import './ProfileModule.js';
// import '../DatabaseModule';
// import { ReactHtmlParser } from 'react-html-parser';
const database = require('../DatabaseModule');
// const profileModules = require('./ProfileModule');

// get comments for a post
const getPostsComments = async (posts) => {
  // iterate through each post, requesting its comments. Wait for all promises to resolve.
  const commentPromiseContainer = [];
  for (let i = 0; i < posts.length; i += 1) {
    // need to implement a get comments function first
    const postComments = database.sendGetRequest('http://localhost:8080/comment/', { id: posts[i].post_id });
    commentPromiseContainer.push(postComments);
  }
  const commentFulfilledContainer = await Promise.all(commentPromiseContainer);

  // assign the comments to each post in the array to hand back
  const postsArray = posts;
  for (let i = 0; i < commentFulfilledContainer.length; i += 1) {
    postsArray[i].comments = commentFulfilledContainer[i];
  }
  // return the array of posts which now each have the additional 'comments' property
  // comments property is itself an array of comment objects
  return postsArray;
};

const getProfile = async (userId) => {
  const profile = await database.sendGetRequest('http://localhost:8080/profile/', { id: userId });
  return profile;
};

const getNamesFromDB = async (postsToSet) => {
  // set up array of unique userIds for which to retrieve names from Db
  const userIdsToRetrieve = [];

  console.log('in get names, postsToSet are: ', postsToSet);

  // loop through and get the userIds of the comment-makers
  for (let i = 0; i < postsToSet.length; i += 1) {
    // console.log('in get outerloop names, postsToSet[i].comments are: ', postsToSet[i].comments);
    for (let j = 0; j < postsToSet[i].comments.length; j += 1) {
      if (userIdsToRetrieve.indexOf(postsToSet[i].comments[j].user_id) === -1) {
        userIdsToRetrieve.push(postsToSet[i].comments[j].user_id);
        // console.log('pushing ', postsToSet[i].comments[j].user_id, 'to userIdsToRetrieve');
      }
    }
  }

  // get the names associated with all the userIds needed on the page
  // console.log('about to go populate promise container with ', userIdsToRetrieve);
  const profilePromiseContainer = [];
  for (let i = 0; i < userIdsToRetrieve.length; i += 1) {
    const commenterProfile = getProfile(userIdsToRetrieve[i]);
    // console.log(' about to push ', commenterProfile, ' onto the profilePromiseContainer');
    profilePromiseContainer.push(commenterProfile);
  }

  console.log('profilePromiseContainer is ', profilePromiseContainer);
  const profileFulfilledContainer = await Promise.all(profilePromiseContainer);
  console.log('profileFulfilledContainer is ', profileFulfilledContainer);

  // taking one level off the array - not sure why it's got an extra nesting?
  const profilesToReturn = [];
  for (let i = 0; i < profileFulfilledContainer.length; i += 1) {
    profilesToReturn.push(profileFulfilledContainer[i][0][0]);
  }

  // attach the names to each comment
  const postsToReturn = postsToSet;
  console.log('profiles to return used to iteratively assign to posts to return is: ', profilesToReturn);
  console.log('posts and comments to iterate through: ', postsToReturn);
  console.log('i will iterate this number of times: ', postsToReturn.length);
  for (let i = 0; i < postsToReturn.length; i += 1) {
    console.log('For i of ', i, ' will iterate ', postsToReturn[i].comments.length, ' times');
    for (let j = 0; j < postsToReturn[i].comments.length; j += 1) {
      // const index = profilesToReturn.userId.indexOf(postsToReturn[i][j].user_id);
      const idToFind = postsToReturn[i].comments[j].user_id;
      const index = profilesToReturn.map((profile) => (profile.user_id)).indexOf(idToFind);
      const firstName = profilesToReturn[index].first_name;
      const lastName = profilesToReturn[index].last_name;
      console.log(firstName, lastName);
      postsToReturn[i].comments[j].name = `${firstName} ${lastName}`;
      // console.log('just assigned sally at postsToReturn[i].comments[j] of
      // ', postsToReturn[i].comments[j]);
    }
  }

  console.log('posts inc profiles ToReturn is ', postsToReturn);
  return postsToReturn;
};

// get posts to display and their associated comments
const getUserPosts = async (userId) => {
  // get posts
  const userPosts = await database.sendGetRequest('http://localhost:8080/post/', { id: userId });

  // get comments
  const postsArray = await getPostsComments(userPosts);

  // get / attach the names to each comment
  console.log('about to hop into getNamesFromDB with postsArray of: ', postsArray);
  const postsArrayWithNames = await getNamesFromDB(postsArray);

  return postsArrayWithNames;
};

function ProfilePage(props) {
  // dummy profile to help implement component frontend with right schema
  // define states
  const [userProfile, setUserProfile] = useState({ biography: '', username: 'Stacy Shapiro' });
  const [userPosts, setUserPosts] = useState([]);

  useEffect(async () => {
    // extract id from props
    console.log('props is: ', props);
    const { userId } = props.state;

    const dummyId = 9;
    console.log(userId);

    // call backend for content linked to userId
    const postsToSet = await getUserPosts(dummyId);
    const profileToSet = await getProfile(dummyId);

    // update state
    setUserPosts(postsToSet);
    setUserProfile(profileToSet);
  }, []);

  useEffect(async () => {
    console.log('state updated, userPosts is now: ', userPosts);
  }, [userPosts]);

  return (
    <div className="App">
      <div id="cover_space">
        <img id="user_profile_pic" src="Stacy meditating.jpg" alt="" />
        <div id="main_title">
          {userProfile.username}
          &apos;s FaceTok Page
        </div>
      </div>
      <div id="nav_button_container">
        <div className="nav_button">Friends</div>
        <div className="nav_button">Photos</div>
        <div className="nav_button">Groups</div>
        <div className="nav_button">Update bio</div>
        <div className="nav_button">Settings</div>
      </div>
      <div id="main_content_container">
        <div id="key_bio_info">
          {userProfile.biography}
        </div>
        <div id="post_container">
          {userPosts.map((post) => (
            <div className="wall_post">
              <img className="wall_post_pic" src={post.photourl} alt="" />
              <p />
              <div className="post_caption">
                {post.caption}
              </div>
              <div className="post_content_container">
                { post.comments.map((comment) => (
                  <div>
                    <p />
                    <b className="post_content_name">{comment.name}</b>
                    <div>
                      {comment.comment_txt}
                    </div>
                  </div>
                )) }
                <p />
                <div className="reply_container">
                  <div className="post_reply_textbox"> Type your reply here...</div>
                  <div className="post_reply_button">Reply!</div>
                </div>
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
}

export default ProfilePage;
