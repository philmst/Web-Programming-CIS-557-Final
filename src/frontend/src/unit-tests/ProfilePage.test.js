import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import ProfilePage from '../ProfilePage';

/**
 * @jest-environment jsdom
 */

const changeLink = jest.fn();

describe('fetch and DOM testing with mocking', () => {
  test('Profile page renders correctly', () => {
    const component = renderer.create(<ProfilePage />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
