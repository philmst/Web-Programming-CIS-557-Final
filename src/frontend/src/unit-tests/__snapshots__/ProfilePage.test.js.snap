// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Profile Page UI Profile page renders correctly 1`] = `
<div
  className="App"
>
  <div
    id="cover_space"
  >
    <img
      alt=""
      id="user_profile_pic"
      src="https://www.psypost.org/wp-content/uploads/2017/08/meditating-woman.jpg"
    />
    <div
      id="main_title"
    >
      Stacy Shapiro
      's FaceTok Page
    </div>
  </div>
  <div
    id="nav_button_container"
  >
    <div
      className="nav_button"
    >
      Friends
    </div>
    <div
      className="nav_button"
    >
      Photos
    </div>
    <div
      className="nav_button"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="link"
      tabIndex={0}
    >
      Groups [active link] 
    </div>
    <div
      className="nav_button"
    >
      Update bio
    </div>
    <div
      className="nav_button"
    >
      Settings
    </div>
  </div>
  <div
    id="main_content_container"
  >
    <div
      id="key_bio_info"
    >
      <b>
        Stacy
        's bio
      </b>
      <p />
      
    </div>
    <div
      id="post_container"
    />
  </div>
</div>
`;
